### ISSUE TITLE
[api-extractor] A package cannot have two exports with the name "Date"

### ISSUE THREAD

**Issue by acalvo** [microsoft/web-build-tools#1350]
See the following minimal repo: https://github.com/acalvo/api-extractor-date

If you `npm install && npx api-extractor run` over it, you get the following error:
```
Error: Internal Error: A package cannot have two exports with the name "Date"
You have encountered a software defect. Please consider reporting the issue to the maintainers of this application.
```
If you change the class name in `date.d.ts` to something different (ex. `Date2`), everything goes OK.

This is a bug, isn't it?

**Comment by octogonz** [comment]
Yes, API Extractor is complaining about the runtime `Date` class conflicting with a locally defined `Date` symbol.  This is technically legal in JavaScript; the local symbol should take precedence over the global symbol.  

However, in doing so, I believe the global symbol becomes impossible to access within that scope.  This will cause trouble for the .d.ts rollup feature. Consider this example:

**file1.ts**
```ts
export class Date { }
```
**file2.ts**
```ts
export function getDate(): Date {
  return new Date();
}
```
**index.ts**
```ts
export { Date } from './file1`';
export { getDate } from './file2`';
```

If we naively rolled these together, the `getDate()` return type would be get confused with the global `Date`.  Instead, API Extractor would need to rename the local declaration everywhere and then unrename it during export, like this:

**rollup.d.ts**
```ts
declare class Date_1 {
}

export { Date_1 as Date };

declare function getDate(): Date;
export { getDate };
```

Consumers will encounter the same exact problem if they import both things.  They would need to rename it, too:
**consumer.ts**
```ts
import { Date as Date_1, getDate } from 'the-lib';
const globalDate: Date = getDate();
```

This is strong evidence that it's a somewhat unwise API design practice to eclipse the global namespace.  It only makes sense if you're /sure/ nobody who wants to consume your API would ever want to use the global API, too.

Just curious -- why did you name your class `Date`?

**Comment by acalvo** [comment]
Thanks for the prompt and detailed answer! :)

> Just curious -- why did you name your class Date?

We define the classes in a language-agnostic format ([protobuf](https://github.com/protocolbuffers/protobuf)) and then transform them to a variety of languages, so we don't consider language-specific matters like this one...

**Comment by bgotink** [comment]
I'm running into this issue as well.

If I remove the `Date` export of the package's entrypoint, I notice the class already gets renamed to `Date_2`. If that could just get renamed in an export statement to `Date` that'd be perfect.

Our usecase, in case you're wondering: we've got a bunch of immutable wrappers around types of values, e.g. `Date`, `EmailAddress`, `Iban`. These can only have valid values, unlike the native `Date` type. This allows functions to e.g. simply request an `Iban` instance as parameter without having to worry about validating that the input is a valid IBAN account number.

**Comment by octogonz** [comment]
> If I remove the `Date` export of the package's entrypoint, I notice the class already gets renamed to `Date_2`. If that could just get renamed in an export statement to `Date` that'd be perfect.

Ha, I think you're right!  I was being dumb about this.  It turns out we already halfway implemented this for the [ambientNameConflict](https://github.com/microsoft/web-build-tools/tree/master/build-tests/api-extractor-scenarios/src/ambientNameConflict) test scenario.

**Comment by octogonz** [comment]
Here's a PR: https://github.com/microsoft/web-build-tools/pull/1400

**Comment by octogonz** [comment]
I confirmed that this fixes @acalvo's repro also.

**Comment by ahutchings** [comment]
@octogonz would you mind publishing a new version of api-extractor with your fix from #1400? I am using the latest version of @microsoft/api-extractor (7.4.1) but it doesn't appear to contain the change.

**Comment by octogonz** [comment]
Ermm... That is really strange. I'll take a look.

**Comment by octogonz** [comment]
Ohhh... it looks like this PR was accidentally closed instead of merged. 😇

@ahutchings  Thanks very much for pointing this out!  I'll reopen PR #1400 and get this fixed.

**Comment by ahutchings** [comment]
Thanks @octogonz!

**Comment by octogonz** [comment]
(Released in 7.4.2)
